From: Simon McVittie <smcv@collabora.com>
Date: Fri, 10 Apr 2026 10:07:14 +0100
Subject: run, context: Mark fd arguments as close-on-exec

On entry to `flatpak run`, these fds have been inheritable (not
FD_CLOEXEC), otherwise they would not have been inherited; but we don't
want the "payload" command to inherit them, so set them as
non-close-on-exec as soon as we receive them. In the cases where we pass
them down to the underlying bwrap command, we'll either dup them, or
set them to be inheritable again (in practice we dup them).

In particular, Chromium-derived web browsers get very upset when their
subsandbox processes inherit unexpected fds, which has been causing crashes
with no useful diagnostic information since CVE-2026-34078 was fixed.

Fixes: 1b5e886d "run: Add --usr-fd and --app-fd options"
Fixes: b5ae89ed "run: Add --(ro-)bind-fd options"
Signed-off-by: Simon McVittie <smcv@collabora.com>
Bug: https://github.com/flatpak/flatpak/issues/6582
Bug-Debian: https://bugs.debian.org/1132968
Forwarded: https://github.com/flatpak/flatpak/pull/6589
---
 app/flatpak-builtins-run.c | 15 +++++++++++++++
 common/flatpak-context.c   |  8 ++++++++
 2 files changed, 23 insertions(+)

diff --git a/app/flatpak-builtins-run.c b/app/flatpak-builtins-run.c
index 23b9f36..48968fb 100644
--- a/app/flatpak-builtins-run.c
+++ b/app/flatpak-builtins-run.c
@@ -81,6 +81,9 @@ option_bind_fd_cb (const char  *option_name,
   if (fd < 3)
     return glnx_throw (error, "File descriptors 0, 1, 2 are reserved");
 
+  if (!flatpak_set_cloexec (fd))
+    return glnx_throw_errno_prefix (error, "--bind-fd");
+
   g_array_append_val (opt_bind_fds, fd);
   fd = -1; /* ownership transferred to GArray */
   return TRUE;
@@ -101,6 +104,9 @@ option_ro_bind_fd_cb (const char  *option_name,
   if (fd < 3)
     return glnx_throw (error, "File descriptors 0, 1, 2 are reserved");
 
+  if (!flatpak_set_cloexec (fd))
+    return glnx_throw_errno_prefix (error, "--ro-bind-fd");
+
   g_array_append_val (opt_ro_bind_fds, fd);
   fd = -1; /* ownership transferred to GArray */
   return TRUE;
@@ -121,6 +127,9 @@ opt_instance_id_fd_cb (const char  *option_name,
   if (fd < 3)
     return glnx_throw (error, "File descriptors 0, 1, 2 are reserved");
 
+  if (!flatpak_set_cloexec (fd))
+    return glnx_throw_errno_prefix (error, "--instance-id-fd");
+
   opt_instance_id_fd = g_steal_fd (&fd);
   return TRUE;
 }
@@ -140,6 +149,9 @@ opt_app_fd_cb (const char  *option_name,
   if (fd < 3)
     return glnx_throw (error, "File descriptors 0, 1, 2 are reserved");
 
+  if (!flatpak_set_cloexec (fd))
+    return glnx_throw_errno_prefix (error, "--app-fd");
+
   opt_app_fd = g_steal_fd (&fd);
   return TRUE;
 }
@@ -159,6 +171,9 @@ opt_usr_fd_cb (const char  *option_name,
   if (fd < 3)
     return glnx_throw (error, "File descriptors 0, 1, 2 are reserved");
 
+  if (!flatpak_set_cloexec (fd))
+    return glnx_throw_errno_prefix (error, "--usr-fd");
+
   opt_usr_fd = g_steal_fd (&fd);
   return TRUE;
 }
diff --git a/common/flatpak-context.c b/common/flatpak-context.c
index b5206aa..5c66a9e 100644
--- a/common/flatpak-context.c
+++ b/common/flatpak-context.c
@@ -1391,6 +1391,14 @@ option_env_fd_cb (const gchar *option_name,
   if (fd < 3)
     return glnx_throw (error, "File descriptors 0, 1, 2 are reserved");
 
+  /* This is not strictly necessary, because we're going to close it after
+   * parsing the environment block, but let's be consistent with other fd
+   * arguments that we need to avoid being inherited by the "payload"
+   * command. This is also a convenient place to validate that it's an
+   * open fd. */
+  if (!flatpak_set_cloexec (fd))
+    return glnx_throw_errno_prefix (error, "--env-fd");
+
   return flatpak_context_parse_env_fd (context, fd, error);
 }
 
